src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClientX/IO/TdsWriteStream.cs

```diff
@@ -0,0 +1,274 @@
+        public void SetPacketSize(int bufferSize) => _WriteBuffer = new byte[bufferSize];
+        
+        public void ReplaceUnderlyingStream(Stream stream)
+            => _underlyingStream = stream;
+
+        /// <inheritdoc />
+        public override bool CanRead => false;
+
+        /// <inheritdoc />
+        public override bool CanSeek => false;
+
+        /// <summary>
+        /// Always returns true since we can always write to the stream.
+        /// </summary>
+        public override bool CanWrite => true;
+
+        /// <inheritdoc />
+        public override long Length => throw new NotSupportedException();
+
+        /// <inheritdoc />
+        public override long Position { get => throw new NotSupportedException(); set => throw new NotSupportedException(); }
+
+        /// <inheritdoc />
+        public override async Task FlushAsync(CancellationToken ct)
+        {
+            await FlushAsync(ct, true).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Use hard flush when we definitely know that the end of packet is reached.
+        /// </summary>
+        /// <param name="ct">Cancellation token for async operation.</param>
+        /// <param name="isAsync">Bool to indicate async operation</param>
+        /// <param name="flushMode">Whether this is a hard flush or a softflush</param>
+        private async Task FlushAsync(CancellationToken ct, bool isAsync, FlushMode flushMode = FlushMode.SoftFlush)
+        {
+            Debug.Assert(PacketHeaderType != TdsStreamPacketType.None, "PacketHeaderType is not set. Cannot flush the buffer without setting the packet header type.");
+            _WriteBuffer[0] = (byte)PacketHeaderType;
+            byte status;
+
+            // TODO: Handle cancellation queueing. If there is a cancellation queued up, then send the status with IGNORE bit set.
+
+            if (flushMode == FlushMode.HardFlush)
+            {
+                status = TdsEnums.ST_EOM;
+            }
+            else
+            {
+                status = TdsEnums.ST_BATCH;
+                // We have filled up the buffer. We need to send it out, hence next packet gets an increment.
+                PacketNumber++;
+            }
+
+            _WriteBuffer[1] = status;
+            _WriteBuffer[2] = (byte)(WriteBufferOffset >> 8);   // Length upper byte
+            _WriteBuffer[3] = (byte)(WriteBufferOffset & 0xff); // Length lower byte
+            _WriteBuffer[6] = PacketNumber;
+            _WriteBuffer[4] = 0;
+            _WriteBuffer[5] = 0;
+
+            _WriteBuffer[7] = 0;
+
+            if (isAsync)
+            {
+                await _underlyingStream.WriteAsync(_WriteBuffer, 0, WriteBufferOffset, ct).ConfigureAwait(false);
+                await _underlyingStream.FlushAsync(ct).ConfigureAwait(false);
+            }
+            else
+            {
+                _underlyingStream.Write(_WriteBuffer, 0, WriteBufferOffset);
```

Comment:
This and next method have cancellation token support

Reply:
For sync writes ? The underlying will be SSL / Network stream.

Reply:
Yep, you're right. I was just going through resharper hints and didn't really think about whether it was applicable here. I think it's suggestions is "you're in an async context, why would you even want to use a synchronous method?" ... which I still don't really understand the why behind this pattern...